be/src/exec/vectorized/partition/partition_hash_map.h

```diff
@@ -0,0 +1,250 @@
+    int32_t remain_size = 0;
+};
+
+// =====================
+// one level partition hash map
+template <PhmapSeed seed>
+using Int8PartitionHashMap = phmap::flat_hash_map<int8_t, PartitionChunks, StdHashWithSeed<int8_t, seed>>;
+template <PhmapSeed seed>
+using Int16PartitionHashMap = phmap::flat_hash_map<int16_t, PartitionChunks, StdHashWithSeed<int16_t, seed>>;
+template <PhmapSeed seed>
+using Int32PartitionHashMap = phmap::flat_hash_map<int32_t, PartitionChunks, StdHashWithSeed<int32_t, seed>>;
+template <PhmapSeed seed>
+using Int64PartitionHashMap = phmap::flat_hash_map<int64_t, PartitionChunks, StdHashWithSeed<int64_t, seed>>;
+template <PhmapSeed seed>
+using Int128PartitionHashMap = phmap::flat_hash_map<int128_t, PartitionChunks, Hash128WithSeed<seed>>;
+template <PhmapSeed seed>
+using DatePartitionHashMap = phmap::flat_hash_map<DateValue, PartitionChunks, StdHashWithSeed<DateValue, seed>>;
+template <PhmapSeed seed>
+using TimeStampPartitionHashMap =
+        phmap::flat_hash_map<TimestampValue, PartitionChunks, StdHashWithSeed<TimestampValue, seed>>;
+template <PhmapSeed seed>
+using SlicePartitionHashMap = phmap::flat_hash_map<Slice, PartitionChunks, SliceHashWithSeed<seed>, SliceEqual>;
+
+// ==================
+// one level fixed size slice hash map
+template <PhmapSeed seed>
+using FixedSize4SlicePartitionHashMap =
+        phmap::flat_hash_map<SliceKey4, PartitionChunks, FixedSizeSliceKeyHash<SliceKey4, seed>>;
+template <PhmapSeed seed>
+using FixedSize8SlicePartitionHashMap =
+        phmap::flat_hash_map<SliceKey8, PartitionChunks, FixedSizeSliceKeyHash<SliceKey8, seed>>;
+template <PhmapSeed seed>
+using FixedSize16SlicePartitionHashMap =
+        phmap::flat_hash_map<SliceKey16, PartitionChunks, FixedSizeSliceKeyHash<SliceKey16, seed>>;
+
+struct PartitionHashMapBase {
+    const int32_t chunk_size;
+
+    PartitionHashMapBase(int32_t chunk_size) : chunk_size(chunk_size) {}
+
+protected:
+    void alloc_new_buffer(PartitionChunks& value, const ChunkPtr& chunk) {
+        value.chunks.push_back(chunk->clone_empty());
+        value.select_indexes.clear();
+        value.select_indexes.reserve(chunk_size);
+        value.remain_size = chunk_size;
+    }
+
+    void flush(PartitionChunks& value, const ChunkPtr& chunk) {
+        if (!value.chunks.empty() && !value.select_indexes.empty()) {
+            value.chunks.back()->append_selective(*chunk, value.select_indexes.data(), 0, value.select_indexes.size());
+            value.select_indexes.clear();
+            value.select_indexes.reserve(chunk_size);
+            value.remain_size = chunk_size - value.chunks.back()->num_rows();
+        }
+    }
+
+    template <typename HashMap, typename KeyLoader>
+    void append_chunk_for_one_key(HashMap& hash_map, ChunkPtr chunk, KeyLoader&& key_loader) {
+        const auto size = chunk->num_rows();
+        for (uint32_t i = 0; i < size; i++) {
+            const auto& key = key_loader(i);
+            auto& value = hash_map[key];
```

Comment:
one chunk for one key?

Reply:
If there are few entries corresponding to a Key, the memory expansion is very serious

Reply:
In this bad case, each entry hold a chunk that only contains one row. And I remove all the `vector::reserve`